doc/newfeat.texi

```diff
@@ -28,11 +22,35 @@ from capturing the cursor.
 @item @ESS{[R]}: New command @code{ess-r-package-use-dir}. It sets the
 working directory of the current process to the current package directory.
 
-@item @ESS{[R]}: Launching or reloading R from a package file now sets
-the working directory to the package path. You can now launch R from
-files within @code{R/} or @code{src/} without worrying about the wd.
+@item @ESS{[R]}: Launching R from a package file now recognises the package
+root as the likely startup directory for R. If
+@code{ess-ask-for-ess-directory} is non-nil, ESS prompts with the
+package root as default selection. Otherwise the working directory is
+automatically set to the package root.
+
+The mechanism is a bit more general if you are using Emacs 25.1 or
```

Comment:
Do we really need this behavior? If I start R in a subdirectory of a git project I probably want to start in that sub-directory not in the root. 

This is an unlikely scenario though and I can leave with it if it indeed simplifies and modernizes the codebase. 

What about "test" sub-directory?

Reply:
I think ESS badly needs a more integrated notion of project and this is the start of it. For data analysis your scripts should work from the root of the project not in subdirectories.

As for the test directory, maybe we can define a notion of subprojects within R packages? Could you describe the use cases for this please? I'm not sure when this would be useful, I never switch to the `test` folder in my workflow. In fact switching to that folder would be disruptive in my case. If we add this feature it should be optional.

> This is an unlikely scenario though

It happens all to the time in my workflow because I often switch to a test file with projectile and then open R. Or I switch to a file in `/R` or `/src`. Or `/.metadata` (which is a nice folder to put your package notes and scratch files because it is automatically ignored by `R CMD build` and you can add that folder in your global `.gitignore` too). I almost never open a file at the root and then open R.

Reply:
> the use cases for this please? 

The use case is as discussed earlier. Tests are evaluated (by R CMD check) in the test sub-directory not at the root. So if I start a fresh R session in test sub-dir, it probably shouldn't be at the root. For the eval part, tests should not be evaluate within the package.

> In fact switching to that folder would be disruptive in my case.

When my test fails, I jump to that particular test and try to figure it out locally. I find it hard to imagine a different workflow. 

> This is an unlikely scenario though

By this I meant that you open R script in a non-R-package.  My earlier concern were exclusively about such case.

Reply:
>  Tests are evaluated (by R CMD check) in the test sub-directory not at the root.

You're right, it might be a good idea to add a notion of subproject later on then. Maybe a better integration of R CMD check would be better, i.e. ideally it would run in a fresh process. Though that poses the question of simulating an installed package which is easiest with `load_all()` but this would impose devtools on users. I personally (almost) never execute these scripts manually because I use testthat and `devtools::test()`.

> When my test fails, I jump to that particular test and try to figure it out locally. I find it hard to imagine a different workflow.

To debug the test you typically don't need R to be in a particular folder, unless the test uses a local resource. I guess for the case of local resources (e.g. a csv file) it does make sense to be in a non-root folder. However it seems it would be hard for ESS to predict which folder because that is dependent on development tools and test scripts.

On the other hand being in a non-root folder makes it difficult to use tools provided by packages like `usethis` or `revdepcheck`.

> By this I meant that you open R script in a non-R-package.

Ok. I think the notion of projects should be helpful to data analysis users. It seems to be good practice for scripts to execute relative to the project root.

Reply:
> I guess for the case of local resources

Precisely.

> However it seems it would be hard for ESS to predict which folder because that is dependent on development tools and test scripts.

Let's just make an exception for "test" sub-directory and be done with it. I guess, we don't want to get too complicated on this. It's not an essential functionality anyways.

> It seems to be good practice for scripts to execute relative to the project root.

Good point. Let's do a bit of nudging then.